In [2]:
import pandas as pd
anno = pd.read_csv(
    "/Users/duc/.nutstore/.nutstore/Resources/Strains_Plasmids/MBT84/Streptomyces_sp_MBT84flye_O_proteins.fa.pantherScore2.2.panther15.0.out",
    sep='\t',
    header = None
)
anno.head()

0                1  \
0  MBT84flye_00140   PTHR42718:SF13   
1  MBT84flye_11305    PTHR35526:SF3   
2  MBT84flye_09875    PTHR15032:SF4   
3  MBT84flye_33120  PTHR12526:SF573   
4  MBT84flye_10235        PTHR47235   

                                                   2              3      4  \
0            METHYL VIOLOGEN RESISTANCE PROTEIN SMVA  2.500000e-156  520.5   
1                         REGULATOR OF SIG15-RELATED   2.300000e-37  127.3   
2  N-ACYL-PHOSPHATIDYLETHANOLAMINE-HYDROLYZING PH...   1.000000e-91  306.8   
3  LIPOPOLYSACCHARIDE CORE BIOSYNTHESIS PROTEIN RFAG   1.000000e-42  145.9   
4                                    BLR6548 PROTEIN   8.600000e-66  221.7   

         5  
0    3-504  
1   21-147  
2   99-355  
3  102-374  
4   10-396

In [4]:
anno.sort_values(3)

0                1  \
5575  MBT84flye_33315    PTHR32294:SF4   
6073  MBT84flye_24880        PTHR43253   
100   MBT84flye_27025  PTHR24221:SF544   
221   MBT84flye_24000   PTHR22855:SF31   
4840  MBT84flye_37075        PTHR43743   
...               ...              ...   
4554  MBT84flye_14050    PTHR21064:SF4   
3354  MBT84flye_00820   PTHR31356:SF48   
6869  MBT84flye_34580   PTHR23500:SF38   
2568  MBT84flye_41660  PTHR11675:SF101   
1312  MBT84flye_16900  PTHR12411:SF360   

                                                      2        3       4  \
5575                         ERROR-PRONE DNA POLYMERASE  0.00000  1338.5   
6073                 TRICORN PROTEASE HOMOLOG 2-RELATED  0.00000  1200.2   
100   MULTIDRUG RESISTANCE-LIKE ATP-BINDING PROTEIN ...  0.00000  1234.0   
221   ACETYL-/PROPIONYL-COA CARBOXYLASE (BETA SUBUNI...  0.00000  1135.3   
4840  POTASSIUM-TRANSPORTING ATPASE ATP-BINDING SUBUNIT  0.00000  1167.4   
...                                                 ...      ...     ...   
4554                                  HOMOSERINE KINASE  0.00099    17.3   
3354    L-ASCORBATE PEROXIDASE 7, CHLOROPLASTIC-RELATED  0.00100    17.3   
6869           TRANSPORTER, PUTATIVE, EXPRESSED-RELATED  0.00100    17.0   
2568    POLYPEPTIDE N-ACETYLGALACTOSAMINYLTRANSFERASE 1  0.00100    16.7   
1312                               OS06G0582600 PROTEIN  0.00100    17.4   

            5  
5575   3-1097  
6073  24-1064  
100   38-1241  
221     1-532  
4840   38-708  
...       ...  
4554   78-199  
3354   20-152  
6869   30-107  
2568  267-337  
1312   26-105  

[7231 rows x 6 columns]

Max e-value = 0.001  
At least 100 aa overlap  
Total annotated = 7231  
Total in database = 9713  
Longer than 100 = 8533

In [9]:
from Bio import SeqIO
proteome = SeqIO.parse("/Users/duc/.nutstore/.nutstore/Resources/Strains_Plasmids/MBT84/Streptomyces_sp_MBT84flye_O_proteins.fa", "fasta")

In [10]:
pl = []
for p in proteome:
    pl.append(len(p))
print(len(pl))
pl[:5]

9713


[93, 208, 117, 575, 231]

In [16]:
th = 100
longpl = [l for l in pl if l >= th]
len(longpl)

8533

In [19]:
anno.loc[:,[0,1]].to_csv(
    '/Users/duc/.nutstore/.nutstore/Resources/Strains_Plasmids/MBT84/Streptomyces_sp_MBT84flye_O_proteins.fa.pantherScore2.2.panther15.0.pantherRef.tsv',
    sep='\t',
    index=False,
    header=False
)

PANTHER give:
Error: Arbitrary sequence ID maps to multiple PANTHER IDs.

In [24]:
idtopacc = {}
for i, r in anno.iterrows():
    if r[0] not in idtopacc:
        idtopacc[r[0]] = [r[1],r[3]]
    else:
        print(f"{r[0]}:{r[1]}, {r[3]}")
        print(f"{r[0]}:{idtopacc[r[0]][0]}, {idtopacc[r[0]][1]}")

MBT84flye_37635:PTHR43877:SF2, 1.6e-09
MBT84flye_37635:PTHR43617:SF20, 1.6e-09
MBT84flye_07480:PTHR22726, 0.00015
MBT84flye_07480:PTHR22726:SF1, 0.00015
MBT84flye_20555:PTHR31270:SF1, 2.4e-58
MBT84flye_20555:PTHR31270, 2.4e-58
MBT84flye_39525:PTHR31084, 1.1e-56
MBT84flye_39525:PTHR31084:SF0, 1.1e-56
MBT84flye_28365:PTHR11780:SF10, 1.1999999999999995e-87
MBT84flye_28365:PTHR11780, 1.1999999999999995e-87
MBT84flye_33270:PTHR12468, 1.1000000000000001e-30
MBT84flye_33270:PTHR12468:SF2, 1.1000000000000001e-30
MBT84flye_18420:PTHR10884, 1.1e-50
MBT84flye_18420:PTHR10884:SF14, 1.1e-50
MBT84flye_43230:PTHR43014:SF2, 7.399999999999999e-41
MBT84flye_43230:PTHR43014, 7.399999999999999e-41
MBT84flye_31915:PTHR46460, 2.3e-43
MBT84flye_31915:PTHR10815:SF5, 2.3e-43


In [35]:
idtopacc = {}
for i, r in anno.iterrows():
    if r[0] not in idtopacc:
        idtopacc[r[0]] = [r[1],r[3]]
    else:
        acccurr = r[1]
        ecurr = r[3]
        accstor = idtopacc[r[0]][0]
        estor = idtopacc[r[0]][1]
        print(f"{r[0]}:{acccurr}, {ecurr}")
        print(f"{r[0]}:{accstor}, {estor}")
        baseacccurr = acccurr.split(':')[0]
        baseaccstor = accstor.split(':')[0]
        if baseacccurr == baseaccstor:
            if not (":" in acccurr and ":" in accstor):
                accchos = baseacccurr
            else:
                accchos = accstor # first appearance
        else: # different base accession, choose the none subfamily one, else choose first appearance
            if (int(":" in acccurr) + int(":" in accstor)) % 2 == 0:
                accchos = accstor
            elif ":" not in acccurr:
                accchos = acccurr
            else:
                accchos = accstor
        print(accchos)
        print()

MBT84flye_37635:PTHR43877:SF2, 1.6e-09
MBT84flye_37635:PTHR43617:SF20, 1.6e-09
2
PTHR43617:SF20

MBT84flye_07480:PTHR22726, 0.00015
MBT84flye_07480:PTHR22726:SF1, 0.00015
PTHR22726

MBT84flye_20555:PTHR31270:SF1, 2.4e-58
MBT84flye_20555:PTHR31270, 2.4e-58
PTHR31270

MBT84flye_39525:PTHR31084, 1.1e-56
MBT84flye_39525:PTHR31084:SF0, 1.1e-56
PTHR31084

MBT84flye_28365:PTHR11780:SF10, 1.1999999999999995e-87
MBT84flye_28365:PTHR11780, 1.1999999999999995e-87
PTHR11780

MBT84flye_33270:PTHR12468, 1.1000000000000001e-30
MBT84flye_33270:PTHR12468:SF2, 1.1000000000000001e-30
PTHR12468

MBT84flye_18420:PTHR10884, 1.1e-50
MBT84flye_18420:PTHR10884:SF14, 1.1e-50
PTHR10884

MBT84flye_43230:PTHR43014:SF2, 7.399999999999999e-41
MBT84flye_43230:PTHR43014, 7.399999999999999e-41
PTHR43014

MBT84flye_31915:PTHR46460, 2.3e-43
MBT84flye_31915:PTHR10815:SF5, 2.3e-43
PTHR46460



In [39]:
idtopacc = {}
for i, r in anno.iterrows():
    if r[0] not in idtopacc:
        idtopacc[r[0]] = [r[1],r[3]]
    else:
        acccurr = r[1]
        ecurr = r[3]
        accstor = idtopacc[r[0]][0]
        estor = idtopacc[r[0]][1]
        baseacccurr = acccurr.split(':')[0]
        baseaccstor = accstor.split(':')[0]
        if ecurr == estor:
            if baseacccurr == baseaccstor:
                if not (":" in acccurr and ":" in accstor):
                    accchos = baseacccurr # accession chosen
                else:
                    accchos = accstor # first appearance
            else: # different base accession, choose the none subfamily one, else choose first appearance
                if (int(":" in acccurr) + int(":" in accstor)) % 2 == 0:
                    accchos = accstor
                elif ":" not in acccurr:
                    accchos = acccurr
                else:
                    accchos = accstor
            echos = ecurr # e value chosen
        else:
            if ecurr > estor:
                accchos = acccurr
                echos = ecurr
            else:
                accchos = accstor
                echos = estor
        idtopacc[r[0]] = [accchos, echos]
pd.DataFrame(idtopacc).T

0         1
MBT84flye_00140   PTHR42718:SF13  2.5e-156
MBT84flye_11305    PTHR35526:SF3   2.3e-37
MBT84flye_09875    PTHR15032:SF4     1e-91
MBT84flye_33120  PTHR12526:SF573     1e-42
MBT84flye_10235        PTHR47235   8.6e-66
...                          ...       ...
MBT84flye_48465        PTHR15394     6e-17
MBT84flye_14790   PTHR30582:SF14   5.4e-43
MBT84flye_04550   PTHR32243:SF18   3.1e-97
MBT84flye_48660   PTHR43214:SF24   1.1e-41
MBT84flye_30160        PTHR13184   2.4e-63

[7222 rows x 2 columns]

In [44]:
csv = pd.DataFrame(idtopacc).T[0].to_csv(
    None,
    sep = '\t',
    header=False
)
with open('/Users/duc/.nutstore/.nutstore/Resources/Strains_Plasmids/MBT84/Streptomyces_sp_MBT84flye_O_proteins.fa.pantherScore2.2.panther15.0.pantherRef.tsv', 'w') as fh:
    fh.write(csv[:-1])